Switch to libb64 for base64 encoding/decoding.

de-vri-es opened the pull request:
`libxmlrpc++` contains an unlicensed base64 implementation (see also #1034). This PR replaces it with `libb64`, a public domain base64 implementation [1]. Benchmarks also indicate that it is significantly faster than the old implementation for large input (more details at #1034).

[1]  http://libb64.sourceforge.net/

de-vri-es committed:
Replace base64.h with libb64.

de-vri-es committed:
Add unit test for base64 encoding/decoding.

tfoote commented:
@de-vri-es Thanks for taking the time to track this down and the study of alternatives and put together this replacement.

dirk-thomas commented:
I am sorry for the late response. The patch looks good to me. The new test could use the ROS code style (rather than tabs etc.) but that is not an important thing.

Since the public header `xmlrpcpp/base64.h` this change might break downstream packages. As far as I see it that it intentional. To make it easier for users which run into this I propose to add a new file with that name which contains just a compile time error message (e.g. `#error some text to explain what happened and how to resolve it`).

de-vri-es committed:
Add a base64.h with #error explaining why it is removed.

de-vri-es commented:
I updated the test to follow ROS style (atleast, spaces and curly braces on new lines). Let me know if I should do more on it.

> Since the public header xmlrpcpp/base64.h this change might break downstream packages. As far as I see it that it intentional. To make it easier for users which run into this I propose to add a new file with that name which contains just a compile time error message (e.g. #error some text to explain what happened and how to resolve it).

Indeed, I considered the base64 encoding to be an internal detail. No need to expose it as a public header. Keeping it internal frees us from having to provide API stability. And if people want to do base64 encoding they're better off finding a dedicated library for that than piggybacking on xmlrpc.

But you're right that the old header was public, so an `#error` makes sense. I added one with a reference to this pull request. It can probably be removed again in ROS M or N.

dirk-thomas commented:
Thank you for the quick update!
